"Fix failing checks" wiki page for new contributors (#154629)
flutter/flutter: docs/contributing/testing/Fix-failing-checks.md

![Some checks were not successful](https://github.com/user-attachments/assets/95fd56e9-4839-4944-b9ac-cc45404896a2)

# How to fix a PR's failing checks

<br>

### tree-status

![Tree is currently broken.](https://github.com/user-attachments/assets/b611d540-c4cb-47dc-a27f-bef8709f24ce)

A **tree-status** failure indicates that the main branch has problems

> that should be addressed anything else is merged in.

Reviewer: ```suggestion
that should be addressed before anything else is merged in.
```

Author: oh my goodness 😂

The result of this check usually goes back and forth a few times each day:
once [review requirements](../Tree-hygiene.md#getting-a-code-review) are met
and all other checks are passing, adding the
[**`autosubmit`**](../../infra/Landing-Changes-With-Autosubmit.md) label
signals a bot to merge the PR once the tree-status check succeeds.

<br>

### Google testing

![Google testing](https://github.com/user-attachments/assets/7d1f9a66-b84a-4223-b57d-77b44f205d1c)

A Google testing failure could be a flake ([see below](#flaking)), or it
might be due to changes in the PR.

Google employees can view the test output and give feedback accordingly.
If 2 weeks have gone by and nobody's looked into it, feel free to
[reach out on Discord](../../Chat.md).

<br>

### A bug in the PR

If you've followed the steps for
[setting up the framework dev environment](../../Setting-up-the-Framework-development-environment.md),
most problems will be caught in static analysis, but oftentimes
there are changes that inadvertently break expected behavior.

If a **customer_testing** check is unsuccessful, it's a signal that a test from

> [flutter/packages](https://github.com/flutter/packages/) has failed.

Reviewer: This should be https://github.com/flutter/tests ? The customer tests registry is there. Although, since flutter/packages also runs tests on master, it _could_ break both?

Author: I believe flutter/tests runs the tests in flutter/packages, and my understanding is that Google testing takes care of the other stuff, but it would be nice to double-check that info.

Reviewer: The customer tests are tests, submitted by third parties, against the framework. These help catch breakages.

IIRC super_list is one of those customers. 

See the note on the `flutter/flutter` Github homepage: "A repository for you to submit your application's tests to ensure that breaking changes don't affect your application."

Author: I experienced quite a bit of frustration a while back after putting a lot of work into [a huge PR](https://github.com/flutter/flutter/pull/149963) only to see that **customer_testing** was failing. For most test failures, I can quickly find the test by pasting the name into the VS Code search tab, but even after doing that in both flutter/flutter and flutter/tests there weren't any results. Eventually I realized that the test was located in flutter/packages; my hope in adding this wiki page is to prevent others from having to go through that 😅

I really appreciate the explanation you gave here. I made an effort to incorporate this info into the page; let me know if it should be tweaked further!

![Some checks were not successful](https://github.com/user-attachments/assets/95fd56e9-4839-4944-b9ac-cc45404896a2)

# How to fix a PR's failing checks

<br>

### tree-status

![Tree is currently broken.](https://github.com/user-attachments/assets/b611d540-c4cb-47dc-a27f-bef8709f24ce)

A **tree-status** failure indicates that the main branch has problems
that should be addressed before anything else is merged in.

> The result of this check usually goes back and forth a few times each day,

Author: I meant that the result alternates between "success" and "failure", often multiple times per day.

But after reading it over again, I agree that the wording isn't super clear. I'll try revising this section.

> A **tree-status** failure indicates that the main branch has problems
> that should be addressed before anything else is merged in.
> 
> The result of this check usually goes back and forth a few times each day:
> once [review requirements](../Tree-hygiene.md#getting-a-code-review) are met
> and all other checks are passing, adding the
> [**`autosubmit`**](../../infra/Landing-Changes-With-Autosubmit.md) label
> signals a bot to merge the PR once the tree-status check succeeds.

Reviewer 2: I suggest we split every section into two parts. (The PR currently is almost like this, but I suggest we make it clearer.) The first paragraph explains what it is, and the second paragraph explicitly tells the reader what to do.

For example, for this one:

```suggestion
A **tree-status** failure indicates that the main branch has problems
that should be addressed before anything else is merged in. No matter
whether it's a false positive or a real problem, it is not related with
with your PR. Flutter maintainers are resolving it as soon as possible.

**What to do:** If your PR already meets the [review requirements](../Tree-hygiene.md#getting-a-code-review)
(for example, have enough code reviews and all checks are green except for this one),
you can add the `autosubmit` label, which signals a bot to merge the PR once the tree-status check succeeds.
And until then, sit back and relax.
```

What do you think?

Reviewer 2: Also, does non-hackers have the permission to add labels?

Author: > I suggest we split every section into two parts. (The PR currently is almost like this, but I suggest we make it clearer.) The first paragraph explains what it is, and the second paragraph explicitly tells the reader what to do.

Sounds fantastic!

<br>

> Also, does non-hackers have the permission to add labels?

Nope—before I had commit access, I'd often awkwardly ask "hey… could someone add `autosubmit`?"
There isn't much that someone without commit access can do, aside from [posting a meme](https://github.com/flutter/flutter/pull/143675#issuecomment-1964923848) and hoping for the best.

Reviewer 2: Ok, in that case we might want to change the 2nd paragraph into:

**What to do:** If your PR already meets the [review requirements](../Tree-hygiene.md#getting-a-code-review)
	(for example, have enough code reviews and all checks are green except for this one),
	you can ask the reviewers to add the `autosubmit` label, which signals a bot to merge the PR once the tree-status check succeeds.
	And until then, sit back and relax.

<br>

### Google testing

![Google testing](https://github.com/user-attachments/assets/7d1f9a66-b84a-4223-b57d-77b44f205d1c)

> A Google testing failure could be a flake ([see below](#flaking)), or it
> might be due to changes in the PR (See
> [Understanding Google Testing](../../infra/Understanding-Google-Testing.md)
> for more info).
> 
> Google employees can view the test output and give feedback accordingly.
> If 2 weeks have gone by and nobody's looked into it, feel free to
> [reach out on Discord](../../Chat.md).

Reviewer 2: I added a bit information and changed it to the style I suggested.

```suggestion
A Google testing failure means that your PR breaks some tests in
Google's internal codebase. It might be a flake ([see below](#flaking))
or a real problem caused by your changes. Only a Google employee
can see the detail of the failure.

**What to do:** Tag a Google employee to look into your issue.
Feel free to [reach out on Discord](../../Chat.md), especially if more than 2
weeks have gone by and nobody's looked into it.

See also:
* [Understanding Google Testing](../../infra/Understanding-Google-Testing.md)
```

Also: I think it might scare contributors to look for random Google employees to investigate this issue. Do you think we can tell them exactly "just go to Discord and ask for help in this channel"?

Author: Thank you!

> I think it might scare contributors to look for random Google employees to investigate this issue. Do you think we can tell them exactly "just go to Discord and ask for help in this channel"?

Personally, I think adding a link to [Chat.md](https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md) is probably the best way to do it.

<br>

### ci.yaml validation

![ci.yaml validation](https://github.com/user-attachments/assets/545a55f8-5bde-460f-92dd-9d87788f9fe8)

> In order for each check to run smoothly, the [.ci.yaml](../../../.ci.yaml)
> file needs to stay in sync with the base branch.
> 
> This check failure can be fixed by applying the latest changes from master.\
> (The [Tree hygiene](../Tree-hygiene.md#using-git) page recommends updating
> via rebase, rather than a merge commit.)
> 
> ![Update with rebase](https://github.com/user-attachments/assets/8bacd87f-410a-4a9c-8ad0-075dd05f3eff)

Reviewer 2: Simplified with the style I suggested:

```suggestion
This failure means that [.ci.yaml](../../../.ci.yaml) is out of date with the base branch.
This file defines the checks that are performed on each PR.

**What to do:** You should update your PR with master, preferably **via rebase** ([reasons](../Tree-hygiene.md#using-git)).

![Update with rebase](https://github.com/user-attachments/assets/8bacd87f-410a-4a9c-8ad0-075dd05f3eff)
```

Author: Looks good overall! I'm not sure the extent to which the [style guide](https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-the-passive-voice-recommend-do-not-require-never-say-things-are-simple) applies here, but I'll probably change it to something similar without using "you" or "we".

Reviewer 2: Yeah I thought about it too. Thanks for linking the style guide. I think the guide definitely applies here as well, since it talks about making docs less intimidating.

<br>

> ### A bug in the PR

Reviewer 2: For the remainder of the article, I still think it'd be the best to categorize them by the failure name (because it's really obvious! I like your idea.)

So something like

```
### Other failures
// Move your "View the test output" here as what-to-do

#### Linux customer_testing
// As a special case of "other failures"
```

Author: (see [below response](https://github.com/flutter/flutter/pull/154629#issuecomment-2387467410))

## A bug in the PR

Oftentimes, a change inadvertently breaks expected behavior.\
When this happens, usually the best way to find out what's wrong is to
[**view the test output**](#view-the-test-output).

If a **customer_testing** check is unsuccessful, it's a signal that something in the
[Flutter customer test registry](https://github.com/flutter/tests/) has failed.
This includes [package tests](../../ecosystem/testing/Understanding-Packages-tests.md)
along with other tests from open-source Flutter projects.\
If a pull request requires an update to those external tests, it qualifies as a
[**breaking change**](../Tree-hygiene.md#handling-breaking-changes);

> it's great to avoid those when possible.

Reviewer 3: ```suggestion
please avoid those when possible.
```

![Some checks were not successful](https://github.com/user-attachments/assets/95fd56e9-4839-4944-b9ac-cc45404896a2)

# How to fix a PR's failing checks

<br>

### tree-status

![Tree is currently broken.](https://github.com/user-attachments/assets/b611d540-c4cb-47dc-a27f-bef8709f24ce)

A **tree-status** failure indicates that the main branch has problems
that should be addressed before anything else is merged in.
The result of this check usually goes back and forth a few times each day,
for a variety of reasons.

**What to do:** Once [review requirements](../Tree-hygiene.md#getting-a-code-review)
are met and all other checks are passing, adding the
[**`autosubmit`**](../../infra/Landing-Changes-With-Autosubmit.md) label

> signals a bot to merge the PR once the tree-status check succeeds.

Reviewer 2: Yeah, that's why I'm not directing contributors toward gaining commit access, but directing them to asking for help from someone with commit access. And also I intentionally did not mention asking for help in the Discord channel, since their reviewer must have commit access and should have added the label. (Sometimes we don't add labels because we don't know if the author wants to make more changes, therefore I might wait for the author to ask me to.)

> signals a bot to merge the PR once the tree-status check succeeds.

Author: My aim was to use your suggestion and tweak it to be entirely geared toward new contributors… what I ended up with had diverged quite a bit from your original suggestion, sorry about that!

I think I'll give it another shot.

might be due to changes in the PR (See
[Understanding Google Testing](../../infra/Understanding-Google-Testing.md)
for more info).
Google employees can view the test output and give feedback accordingly.

**What to do:** If 2 weeks have gone by and nobody's looked into it,
feel free to [reach out on Discord](../../Chat.md).

<br>

### ci.yaml validation

![ci.yaml validation](https://github.com/user-attachments/assets/545a55f8-5bde-460f-92dd-9d87788f9fe8)

In order for checks to run correctly, the [.ci.yaml](../../../.ci.yaml)
file needs to stay in sync with the base branch.

**What to do:** This check failure can be fixed by applying the latest changes
from master.\
(The [Tree hygiene](../Tree-hygiene.md#using-git) page recommends updating
via rebase, rather than a merge commit.)

![Update with rebase](https://github.com/user-attachments/assets/8bacd87f-410a-4a9c-8ad0-075dd05f3eff)

<br>

## A bug in the PR

Following the steps outlined in
[setting up the framework dev environment](../../Setting-up-the-Framework-development-environment.md)
makes it so most problems are caught in static analysis, but oftentimes a change
still inadvertently breaks expected behavior.

> 

Author: Hopefully https://github.com/flutter/flutter/pull/154629/commits/8e131e294dbe26c718bde9430c23147e619ebc2d helps make this easier to navigate! :)

This was caught by the test expectation on the following line:
  file:///b/s/w/ir/x/w/flutter/packages/flutter/test/widgets/async_test.dart line 115
The test description was:
  gracefully handles transition from null future
════════════════════════════════════════════════════════════════════════════════════════════════════
```

From there, it's just a matter of finding the failing test,
[running it locally](./Running-and-writing-tests.md),
and figuring out how to fix it!

<br>

### Flaking

A check might "flake", or randomly fail, due to an
[infra error](../../infra/Understanding-a-LUCI-build-failure.md#overview-of-an-infra-failure-build)
or for other reasons unrelated to the PR's changes.

Sometimes a flake will resolve itself once some more changes are pushed to
re-trigger the checks. This can be as simple as [performing a rebase](#ciyaml-validation)
to include the latest changes from the main branch.

If that doesn't work, head over to the [tree status](https://flutter-dashboard.appspot.com/#/build?repo=flutter)
page and check if that test is failing on the main branch too.

- If so: wait for the tree to turn green, and then rebase with those changes.

> - But if not: it's probably [a bug in the PR](#a-bug-in-the-pr).

Author: > removed the word "simple", per the style guide

One would think I'd always have this in mind, after how many times I've read through that section… thanks for catching that! 😅

I ended up rewriting most of this section; hopefully it looks better now!
